### PR TITLE
Bytes crate support

### LINKED ISSUE #4: Add bytes crate support
Add bytes crate support and fix tests

### PR DESCRIPTION
**Pull request by metaclips**
- [x] Project structure
- [x] Add bytes crate support
- [x] Add tests
- [x] Add benchmark

closes #4 and #3

### CODE REVIEW

**Comment by metaclips** [comment]
pr ready @rainliu @Sean-Der

File: Cargo.toml [hunk 1]
```diff
@@ -14,7 +14,12 @@ repository = "https://github.com/webrtc-rs/rtp"
 [dependencies]
 byteorder = "1.3.2"
 rand = "0.8.0"
-thiserror = "1.0.23"
+bytes = "1.0.0"
```
Comment (rainliu) [inline Cargo.toml:17]: Looks like this PR overwrote the recent changes with regards to thiserror related error handling refactoring. Could you still keep that changes?

File: src/codecs/opus/mod.rs [hunk 2]
```diff
@@ -1,38 +1,54 @@
...
 
 impl Depacketizer for OpusPacket {
-    fn depacketize<R: Read>(&mut self, reader: &mut R) -> Result<(), Error> {
-        self.payload.clear();
-        reader.read_to_end(&mut self.payload)?;
-        if self.payload.is_empty() {
-            Err(Error::PayloadIsNotLargeEnough)
-        } else {
-            Ok(())
+    fn unmarshal(&mut self, packet: &mut BytesMut) -> Result<BytesMut, RTPError> {
```
Comment (rainliu) [inline src/codecs/opus/mod.rs:30]: Prefer new thiserror based error handling

File: src/codecs/vp8/vp8_test.rs [hunk 3]
```diff
@@ -1,146 +1,123 @@
...
+        assert_eq!(
+            result.err(),
+            Some(RTPError::ShortPacket),
+            "Result should be err in case of error",
+        );
+
+        // Normal packet
+        let raw_bytes = vec![0x00, 0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x90];
+        let result = pck.unmarshal(&mut raw_bytes[..].into())?;
+        assert!(!result.is_empty(), "Payload must be not empty");
```
Comment (rainliu) [inline src/codecs/vp8/vp8_test.rs:38]: What about those asserts for pick.x/I/l...?

File: src/errors.rs [hunk 4]
```diff
@@ -0,0 +1,63 @@
+#[derive(Debug, Clone, Copy, PartialEq)]
```
Comment (rainliu) [inline src/errors.rs:1]: Keep thiserror based error handling

File: Cargo.toml (hunk 1 above)
Comment (metaclips) [reply Cargo.toml:17]: Will do that

Comment (metaclips) [reply Cargo.toml:17]: @rainliu not all error values follow pion rtp so unneeded will be removed

**Comment by Plecra** [comment]
By the benchmarks added in this PR, the `bytes`-based implementation causes some pretty bad performance regressions:

```
Marshal Benchmark       time:   [178.41 ns 179.35 ns 180.62 ns]
                        change: [+90.552% +95.080% +99.274%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 21 outliers among 100 measurements (21.00%)
  3 (3.00%) high mild
  18 (18.00%) high severe

Marshal_To Benchmark    time:   [101.32 ns 101.54 ns 101.78 ns]
                        change: [+212.08% +227.10% +242.31%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 13 outliers among 100 measurements (13.00%)
  9 (9.00%) high mild
  4 (4.00%) high severe

Shared Struct           time:   [312.47 ns 315.73 ns 320.07 ns]
                        change: [+10.562% +550.50% +1634.9%] (p = 0.22 > 0.05)
                        No change in performance detected.
Found 12 outliers among 100 measurements (12.00%)
  5 (5.00%) low mild
  4 (4.00%) high mild
  3 (3.00%) high severe

New Struct              time:   [405.03 ns 405.68 ns 406.35 ns]
                        change: [+14.377% +15.663% +16.854%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 8 outliers among 100 measurements (8.00%)
  2 (2.00%) high mild
  6 (6.00%) high severe
  ```
[Implementation](https://gist.github.com/Plecra/4beb78c18452ec49a4802e19882a403d)

**Comment by metaclips** [comment]
> By the benchmarks added in this PR, the `bytes`-based implementation causes some pretty bad performance regressions:
> 
> ```
> Marshal Benchmark       time:   [178.41 ns 179.35 ns 180.62 ns]
>                         change: [+90.552% +95.080% +99.274%] (p = 0.00 < 0.05)
>                         Performance has regressed.
> Found 21 outliers among 100 measurements (21.00%)
>   3 (3.00%) high mild
>   18 (18.00%) high severe
> 
> Marshal_To Benchmark    time:   [101.32 ns 101.54 ns 101.78 ns]
>                         change: [+212.08% +227.10% +242.31%] (p = 0.00 < 0.05)
>                         Performance has regressed.
> Found 13 outliers among 100 measurements (13.00%)
>   9 (9.00%) high mild
>   4 (4.00%) high severe
> 
> Shared Struct           time:   [312.47 ns 315.73 ns 320.07 ns]
>                         change: [+10.562% +550.50% +1634.9%] (p = 0.22 > 0.05)
>                         No change in performance detected.
> Found 12 outliers among 100 measurements (12.00%)
>   5 (5.00%) low mild
>   4 (4.00%) high mild
>   3 (3.00%) high severe
> 
> New Struct              time:   [405.03 ns 405.68 ns 406.35 ns]
>                         change: [+14.377% +15.663% +16.854%] (p = 0.00 < 0.05)
>                         Performance has regressed.
> Found 8 outliers among 100 measurements (8.00%)
>   2 (2.00%) high mild
>   6 (6.00%) high severe
> ```
> 
> [Implementation for master](https://gist.github.com/Plecra/4beb78c18452ec49a4802e19882a403d)

Thanks @Plecra will look into it

**Comment by metaclips** [comment]
> By the benchmarks added in this PR, the `bytes`-based implementation causes some pretty bad performance regressions:
> 
> ```
> Marshal Benchmark       time:   [178.41 ns 179.35 ns 180.62 ns]
>                         change: [+90.552% +95.080% +99.274%] (p = 0.00 < 0.05)
>                         Performance has regressed.
> Found 21 outliers among 100 measurements (21.00%)
>   3 (3.00%) high mild
>   18 (18.00%) high severe
> 
> Marshal_To Benchmark    time:   [101.32 ns 101.54 ns 101.78 ns]
>                         change: [+212.08% +227.10% +242.31%] (p = 0.00 < 0.05)
>                         Performance has regressed.
> Found 13 outliers among 100 measurements (13.00%)
>   9 (9.00%) high mild
>   4 (4.00%) high severe
> 
> Shared Struct           time:   [312.47 ns 315.73 ns 320.07 ns]
>                         change: [+10.562% +550.50% +1634.9%] (p = 0.22 > 0.05)
>                         No change in performance detected.
> Found 12 outliers among 100 measurements (12.00%)
>   5 (5.00%) low mild
>   4 (4.00%) high mild
>   3 (3.00%) high severe
> 
> New Struct              time:   [405.03 ns 405.68 ns 406.35 ns]
>                         change: [+14.377% +15.663% +16.854%] (p = 0.00 < 0.05)
>                         Performance has regressed.
> Found 8 outliers among 100 measurements (8.00%)
>   2 (2.00%) high mild
>   6 (6.00%) high severe
> ```
> 
> [Implementation for master](https://gist.github.com/Plecra/4beb78c18452ec49a4802e19882a403d)

@Plecra performance has been improved, do review :-)

**Comment by falconmick** [comment]
Cheers for all the work that has gone into this PR BytesMut is defiantly a better api interface than The current

**Comment by metaclips** [comment]
Haa, thanks @falconmick. There's also work on using byte slice https://github.com/metaclips/rtp/tree/use-byte-slice. Performance seems niceeeeee @rainliu @Plecra  I do plan on taking out time to fix tests and reviews.

**Comment by falconmick** [comment]
I will admit, the idea of a byte slice is more appealing, but I really don’t know enough to provide much feedback

**Comment by metaclips** [comment]
Opening a different PR that using std u8 array

**Pull request closed without merging**